### ISSUE TITLE
ENH: improve fitting of scipy.stats distributions using appropriate bounds and constraints

### ISSUE THREAD

**Issue by Gabriel-Kissin** [scipy/scipy#23515]
### Is your feature request related to a problem? Please describe.

This follows on from a separate issue https://github.com/scipy/scipy/issues/23503 in which this question was touched upon. I thought I'd pull all the relevant comments into one new thread so they don't get lost in that discussion, as there are valuable points scattered over the [issue](https://github.com/scipy/scipy/issues/23503) and additional points in the associated [pull request](https://github.com/scipy/scipy/pull/23504).

The conversation there was about trapezoid fit failing to converge. @nickodell pointed out that the solver's simplex included out-of-bounds values (bounds: `c,d ∈ [0,1]`):
> > Looking at the behavior of the Nelder-Mead solver, it looks like c = 1 d = 1 causes the initial simplex to be the following:
> >
> > ```
> > initial_sim [[1.   1.  ]
> >  [1.05 1.  ]
> >  [1.   1.05]]
> > ```
> @nickodell do you happen to know whether the solver is bounded? I'd've assumed that when minimising negative log likelihood, that it would give the solver bounds for the args. These bounds could be simply taken for every distribution from its `_shape_info`. But from that simplex that doesn't seem to be done?
> 
> (In fact I'd consider giving the solver constraints as well as bounds. For example for trapezoid we must have `c<=d`, which isn't expressed by the bounds (`c,d ∈ [0,1]`) alone. These constraints can be taken using each distribution's `_argcheck` as a constraint function, i.e. constrain the search to `c,d` values for which `_argcheck` returns `True`. However using constraints means limits on which solver algorithms can be used). 

 _Originally posted by @Gabriel-Kissin in [#23503](https://github.com/scipy/scipy/issues/23503#issuecomment-3221501030)_

> I can tell you that the Nelder-Mead optimizer contains code to avoid this case when you provide x0 which is right up against the bounds. (See #20345) However, fit() doesn't appear to supply any bounds to Nelder-Mead, so that code never runs. I couldn't tell you why - I mostly follow scipy.optimize development. Someone with scipy.stats knowledge may be able to tell you.
> 
> Using constraints would pretty significantly reduce your options on which optimizers you could use. Only COBYLA, COBYQA, SLSQP and trust-constr support them.
> 
> Of those options, I would probably not pick COBYLA or SLSQP, as they have a number of very weird numerical behaviors. (e.g in some cases they can minimize the function `f(x)`, but not the function `1000000*f(x)`.) That leaves trust-constr and COBYQA, which I suspect weren't chosen due to the fact that they're quite slow. 

 _Originally posted by @nickodell in [#23503](https://github.com/scipy/scipy/issues/23503#issuecomment-3221529872)_

> > Using constraints would pretty significantly reduce your options on which optimizers you could use. Only COBYLA, COBYQA, SLSQP and trust-constr support them.
> 
> Or `differential_evolution`, which is the slowest of them all. I think `trust-constr` is probably the best to use if there aren't really any local minima, `differential_evolution` is much more likely to find the global minimum.
> 
> `nelder-mead` simply clips values to the bounds. It doesn't [look like](https://github.com/scipy/scipy/blob/main/scipy/stats/_distn_infrastructure.py#L2770) bounds are supplied to the optimizer.
> 
> In that vein, adding in constraints/bounds/etc looks to be a bit of work. 

 _Originally posted by @andyfaff in [#23503](https://github.com/scipy/scipy/issues/23503#issuecomment-3222057608)_

> For trapezoid, setting c = 1 and d = 1 causes at least one value in the initial simplex to be outside the bounds of valid parameters.
> 
> It looks like a similar problem can be found in the default settings for the following other distributions:
> 
> * triang
> * ksone
> * kstwo
> 
> See the following program for how I found this list: https://gist.github.com/nickodell/70f1172e3c667e8d34008933f303870c 

 _Originally posted by @nickodell in [#23503](https://github.com/scipy/scipy/issues/23503#issuecomment-3222194111)_

> @mdhaber, does it make sense to start adding more complexity to the fitting mechanism here? There are several situations where using the following could make sense:
> 
> - one shape parameter should be less than the other (`LinearConstraint`/`NonlinearConstraint`)
> - where shape parameters should remain within limits (`Bounds`)
> 
> Or would effort be better directed towards the new `ContinuousDistribution` infrastructure?

_Originally posted by @andyfaff in https://github.com/scipy/scipy/issues/23504#issuecomment-3226712872_
            

### Describe the solution you'd like.

- Bounds: 
  - the values for the bounds are already known for each distribution, as mentioned, in `_shape_info` 
  - parsing bounds from _shape_info and passing them to the solver should be relatively simple.
  - bounds may be helpful not just for shape params but also for `scale>0`.
- Constraints:
  - E.g. for the trapezoid distribution, we must have c<=d, which isn't expressed by the bounds `c,d ∈ [0,1]` alone
  - This could be given to the solver like `constraints = [{"type": "ineq", "fun": lambda params: params[1] - params[0]}]  # d - c >= 0`
  - To avoid specifying such constraints for each distribution, we could perhaps use the `_argcheck` method each distribution already has (would it matter that `_argcheck` only returns bool True/False instead of a float indicating how far from the constraint we are?)
  - This could be given to the solver like `constraints = [{"type": "ineq", "fun": lambda params: distribution._argcheck(*params)}]  # general for any dist`

I imagine implementing bounds would be easier than implementing constraints. It would also place less limitations on which solvers can be used as discussed. Giving at least bounds does seem like an obvious thing to do

If just bounds were implemented, would nelder-mead still work? Given that its way of doing bounds is simply to clip the simplex. In the trapezoid case, such clipping would result in a simplex like
```
[[1.   1.  ]    # [[1.   1.  ]
 [1.   1.  ]    #  [1.05 1.  ]
 [1.   1.  ]]   #  [1.   1.05]]
```
which wouldn't help the solver much... so perhaps a solver which has proper support for bounds would be necessary

**Comment by andyfaff** [comment]
It's better to use LinearConstraint and NonlinearConstraint than the dictionaries. Also use a Bounds object. 
When you supply bounds and constraints, but don't specify a method, then `minimize` chooses a suitable method automatically.

**Comment by andyfaff** [comment]
_fitstart could create the constraints, if known.
Actually, returning some kind of dict with specific minimize keys could work well

**Comment by Gabriel-Kissin** [comment]
The dict idea sounds good. You mean `_fitstart` should return something like
```python
{
    "x0": initial_guess,
    "bounds": Bounds(...),
    "constraints": [(Non)LinearConstraint(...)],
}
```
(it could also include `method` for which type of solver to use, or any [other `minimize` params](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html))

which can then be fed straight into `minimize` like
```python
options = dist._fitstart(data)
res = minimize(
    fun=negloglike,
    **options   # unpack x0, bounds, constraints, method, etc.
)
```

**Comment by Gabriel-Kissin** [comment]
What I'd had in mind was separate `_fitbounds` and `_fitconstraints` methods. This could be implemented in `rv_continuous` (simply parsing the `_shape_info` and using `_argcheck`) and inherited by all distributions, overwritten if necessary in specific distributions. 

For many distributions we would want `_fitconstraints` to return `()` for no constraints, so that the minimiser has a wider range of appropriate methods to choose from. Presumably `_fitbounds` should always return some bounds (even for distributions with no shape parameters) if only to ensure `scale>0`.

I guess that having three separate methods would allow more flexibility. E.g. `_fitstart` could be overwritten in a particular dist, while leaving `_fitbounds` and `_fitconstraints` at the default inherited version.

(Perhaps the best of both worlds would be to have those three separate methods, and then have a final method `minimiser_kwargs` which just calls all those methods and returns a dict as you suggest - meaning cleaner code where `fit` calls `minimize`.)

**Comment by nickodell** [comment]
@andyfaff

>nelder-mead simply clips values to the bounds.

Isn't it doing something more sophisticated, at least for the initial simplex?

https://github.com/scipy/scipy/blob/dd36f954626d38509482c475de8740a262668449/scipy/optimize/_optimize.py#L832-L842

I tested this, by supplying bounds of [0, 1] for c and d, without changing their initial values, and it does improve the fit. It allows the c parameter to be fit, but d is still stuck at 1. (The initial simplex contains the point c = 1, d = 0.95, which violates the c < d rule of the distribution.)

**Comment by andyfaff** [comment]
Yes, a bit posher for the initial simplex.

**Comment by mdhaber** [comment]
> @mdhaber, does it make sense to start adding more complexity to the fitting mechanism here ... Or would effort be better directed towards the new `ContinuousDistribution` infrastructure?

gh-23503 started with the premise "Following the principle that things should 'just work' out of the box". That's tough for generic, automatic distribution fitting given that optimizers do not guarantee convergence for arbitrary objective functions with arbitrary objectives and constraints from arbitrary guesses!

I've done quite a bit for `stats`'s fitting capabilities over the past few years (e.g. gh-15436, gh-11782, gh-16967, gh-17141, gh-19475...), and I found that it's quite easy to make just as much work out of fitting than all the rest of the distribution infrastructure combined. So recently, I've tried to save SciPy maintainers some trouble by working with [`zfit`](https://zfit.readthedocs.io/en/stable/) to get them to work with our new distribution infrastructure. Consequently, I wouldn't recommend putting time into fitting with either the old or new infrastructures; just let people who have full-time interest in fitting handle fitting (although it looks like that could use a nudge). Some related thoughts are in the [new infrastructure tutorial](https://docs.scipy.org/doc/scipy/tutorial/stats/rv_infrastructure.html#maximum-likelihood-estimation).

---

BTW, it doesn't look like this has been mentioned: `rv_continuous.fit` _does_ enforces constraints and bounds - just with a [penalty method](https://github.com/scipy/scipy/blob/de6e98663d4aa2a3ce7708fd089144cbb2c9d7d2/scipy/stats/_distn_infrastructure.py#L1588). When I started looking into it a few years ago, it seemed so crude, and I figured it should be easy to improve performance by using "real" bounds and constraints. I've played around with it a few times since then - even as recently as drafting a [`fit` method for the new infrastructure](https://github.com/mdhaber/scipy/blob/1033ef4800a07ebdd4e1aa7641450f61596c0f24/scipy/stats/_distribution_infrastructure.py#L4977) - but the cost/benefit ratio of replacing it actually hasn't looked very good to me, so I've decided against including it. I think fitting is super important and interesting - it's just also really hard to make automatic, so I think there are better ways to solve the problem than pouring maintainer time into it.

**Comment by andyfaff** [comment]
Thumbs up for the new infrastructure tutorial, it's well written.

**Comment by Gabriel-Kissin** [comment]
Thanks @mdhaber - that all makes sense and fits with my experiences. I've recently been working on code to fit distributions to binned data (histograms), so even distributions which generally have an analytic closed form solution don't have one here. And neither the distribution's `.fit` method nor the `.nnlf` method work for histograms, so implemented them myself using `minimize` directly. I used bounds taken from `._shape_info`, constraints using `._argcheck`, and x0 using `._fitstart` - and for many distributions it does work painlessly and well. However it often doesn't... 

I think one issue is that the bounds for each shape param given in `_shape_info` are the mathematical/theoretical bounds for the variable, and not the practical bounds. So lots of numerical instability. Clipping the bounds helps, but for some distributions this needs to be so stringent that perfectly reasonable values for other distributions are excluded. So hard to do this in a one-size-fits-all approach.

And thanks for the intro to zfit - it looks like it can [fit to binned data](https://zfit.readthedocs.io/en/stable/_tmp/zfit-tutorials/components/33%20-%20Binned%20fits.html#) but I suspect the new API will have a steep learning curve

**Comment by mdhaber** [comment]
@jonas-eschle any updates on SciPy compatibility? How can we help?

**Comment by jonas-eschle** [comment]
Hi @mdhaber , thanks for tagging, apologies for the late reply (vacation, duty travel,...). I am more responsive next week.

> And thanks for the intro to zfit - it looks like it can [fit to binned data](https://zfit.readthedocs.io/en/stable/_tmp/zfit-tutorials/components/33%20-%20Binned%20fits.html#) but I suspect the new API will have a steep learning curve

@nickodell the API should not change too much, in fact, you can (WIP) create SciPy distributions and should use them with a zfit loss directly. All that is left is the minimization, i.e.

```
loss = zfit.loss.UnbinnedNLL(scipydist, data)
minimizer = zfit.minimize.ScipyBFGS()
result = minimizer.minimize(loss)
```
or of course just 
```
result = zfit.minimize.ScipyBFGS().minimize(loss)
```

There are very good reasons to split this into two calls (see e.g. the stalled PR https://github.com/scipy/scipy/pull/8414).

The goal of zfit is to provide advanced fitting capabilities: pre-defined, optimized cost functions, as you've noticed first-class support for binned PDFs (and mixed), first class support for multidimensional model building etc.

Making this available to work with SciPy, the goal is that for the _simple_cases, not a lot will change, but a lot more functionality will be available. Ideally basically nothing changes. But more complex cases will be possible quite easy, and zfit consumes the SciPy objects while offering a lot of advanced features. That's at least the overall goal.

In terms of API, zfit and the new SciPy dists are quite similar, and only minor things actually disagree (afai can see). Since we're also working on a large overhaul, V2, with possibly some API changes, this would be very welcome. zfit generally offers a more complex API, covering more general use-cases.

You mentioned also the constraints; these are a bit tricky in a way: they can be implemented as penalties (which is easy to do in zfit), or for specific cases (as Matt mentioned). If possible, I found it more efficient to reformulate the problem, but this won't stop certain use-cases.

It would be great to actually have an idea what you would wish, in general, that a fitting library could do more and SciPy is currently lacking. I've created a [discussion on the zfit repo](https://github.com/zfit/zfit/discussions/670), feel free to add anything. Or even have a chat (just drop me an e-mail) to exchange ideas.
